vhdl_syntax/src/parser/productions/names.rs

```diff
@@ -32,4 +74,480 @@ impl<T: TokenStream> Parser<T> {
             self.end_node();
         }
     }
+    pub fn name_list(&mut self) {
+        self.start_node(NameList);
+        self.separated_list(Parser::name, Comma);
+        self.end_node();
+    }
+
+    fn suffix(&mut self) {
+        // LRM §8.3
+        // suffix ::= identifier | string_literal | character_literal | `all` ;
+        self.expect_one_of_tokens([
+            Identifier,
+            StringLiteral,
+            CharacterLiteral,
+            Keyword(Kw::All),
+        ]);
+    }
+
+    fn name_tail(&mut self) {
+        // name             ::= prefix [ name_tail ] ;
+        // name_tail        ::= selected_name | attribute_name | indexed_name | slice_name | function_name ;
+        // selected_name    ::= `.` suffix [ name_tail ] ;
+        // attribute_name   ::= [ signature ] `'` identifier [ `(` expression `)` ] [ name_tail ] ;
+        // function_name    ::= `(` association_list `)` [ name_tail ] ;
+        // indexed_name     ::= `(` expression { `,` expression } `)` [ name_tail ] ;
+        // slice_name       ::= `(` discrete_range `)` [ name_tail ] ;
+
+        if self.next_is(Dot) {
+            self.start_node(SelectedName);
+            self.expect_token(Dot);
+            self.suffix();
+            self.end_node();
+            self.name_tail();
+        } else if self.next_is(LeftPar) {
+            // Try to differentiate between function calls, indexed names and slices as good as possible:
+            // 1. An `association_list` can be uniquely identified by searching for a '=>' inside the parenthesis
+            // 2. When at least a single comma is found, try parse the contents as an expression list
+            // 3. When the `to` or `downto` keyword is found, parse the content as a slice name
+            //
+            // If none of these apply, it can be either a `subtype_indication` or a single `expression`
+
+            if self.lookahead_in_parens([RightArrow]).is_some() {
+                self.start_node(FunctionCallOrIndexedName);
+                self.expect_token(LeftPar);
+                self.association_list();
+                self.expect_token(RightPar);
+                self.end_node();
+            } else if self.lookahead_in_parens([Comma]).is_some() {
+                self.start_node(FunctionCallOrIndexedName);
```

Comment:
This could also be a `SliceName`, if all names are range expression, i.e.
`name(some_type'range, some_other_type(0)'range)`.

It might be worth thinking about extending expressions so that a `range` is a valid expression too (as far as the parser is concerned). This could simplify the grammar a lot and make parsing easier.

Reply:
Is that so? The LRM gives the production of `slice_name` as `slice_name ::= prefix ( discrete_range )`.

Reply:
Ah right, my mistake, sorry